pkg/operator/controller/canary/controller.go

```diff
@@ -0,0 +1,177 @@
+package canary
+
+import (
+	"fmt"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+	"github.com/openshift/cluster-ingress-operator/pkg/manifests"
+
+	"github.com/google/go-cmp/cmp"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "canary_controller"
+)
+
+var (
+	log = logf.Logger.WithName(controllerName)
+)
+
+// New creates the canary controller.
+//
+// The canary controller will watch the Default IngressController, as well as
+// the canary service, daemonset, and route resources.
+func New(mgr manager.Manager, config Config) (controller.Controller, error) {
+	reconciler := &reconciler{
+		Config: config,
+		client: mgr.GetClient(),
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+
+	// Only trigger a reconcile request for the canary controller via events for the default ingress controller.
+	defaultIcPredicate := predicate.NewPredicateFuncs(func(meta metav1.Object, object runtime.Object) bool {
+		return meta.GetName() == manifests.DefaultIngressControllerName
+	})
+
+	if err := c.Watch(&source.Kind{Type: &operatorv1.IngressController{}}, &handler.EnqueueRequestForObject{}, defaultIcPredicate); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Reconcile ensures that the canary controller's resources
+// are in the desired state.
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	result := reconcile.Result{}
+	errors := []error{}
+
+	if err := r.ensureCanaryNamespace(); err != nil {
+		errors = append(errors, err)
+		return result, utilerrors.NewAggregate(errors)
```

Comment:
The returns (generally) could become just:

    return result, utilerrors.NewAggregate(append(errors, err))

Reply:
Good point.

Reply:
fixed, thanks!